test/test_jit.py

```diff
@@ -2179,6 +2179,34 @@ def test_dropout(self):
             m = self.createFunctionFromGraph(g)
             self.assertEqual(outputs, m(*inputs))
 
+    def test_dropout_requires_grad(self):
+        class MyModule(torch.nn.Module):
+            def __init__(self, M):
+                super(MyModule, self).__init__()
+
+                self.dropout = torch.nn.Dropout(0.5)
+                self.linear = torch.nn.Linear(M, M)
+
+            def forward(self, input):
+                input = self.dropout(input)
+                output = self.linear(input)
+                return output
+
+        def func(input):
+            return F.dropout(input, 0.5, training=True)
+
+        M = 1000
+        scripted_module = torch.jit.script(MyModule(M))
+        scripted_func = torch.jit.script(func)
+
+        X = torch.randn(M, M, requires_grad=True)
+        FileCheck().check("aten::bernoulli_").run(scripted_module.graph_for(X))
+        FileCheck().check("aten::bernoulli_").run(scripted_func.graph_for(X))
+
+        X = torch.randn(M, M, requires_grad=False)
```

Comment:
The condition here is a bit weird. It should not depend on whether `x` is requiring grad, but on whether `m` is in training mode, right?

Reply:
The torchscript executor only bothers to split out symbolic gradients when the inputs require_grad, which is probably why these flags are getting set here.

Reply:
My question was more about why there is not a .eval() here to switch the scripted module to eval mode. Unless .graph_for auto selected train / eval mode based on input requires_grad which sounds wrong to me.

Reply:
@SsnL @zdevito @bddppq Thanks for the comments! To make the testcase a bit more clear, I've added a comment here to explain the behavior. https://github.com/pytorch/pytorch/pull/29436/files#diff-78aab5b4217aeaed920efab0c5f5f5c8R2202
Please let me know if that looks good! Thanks!

Reply:
Thank you so much!!!